### ISSUE TITLE
Add ability to efficiently find duplicates in a large dataset to the CLI

### ISSUE THREAD

**Issue by seadowg** [LSHTM-ORK/ODK_Biometrics#50]
Currently, if you have sets of templates (the left and right thumbs for example) from individuals and want to find duplicates within that, you'd need to write a script that compares pairs of templates using `match`. Without a lot of work to deal with this in parallel, this would be very slow as every set has to be compared to each other set.

Keppel's CLI could help with this by providing a command to find matches within a set of data and execute the comparisons in parallel to speed things up. Something like:

```bash
keppel pmatch -i input.csv -o output.csv -t 40 -p 16
```

The input CSV could have columns `id`, `template_1`, `template_2` etc and the output CSV would have `id_1`, `id_2`, `score_1`, `score_2` etc. The -t and -p arguments would allow customising the matching threshold (limiting which pairs get output) and the number of threads used to execute comparisons in parallel respectively.

**Comment by seadowg** [comment]
Very interested in thoughts from @chrissyhroberts here! @tobiasmcnulty has already built a working version of this for a specific use case (https://github.com/caktus/kafis), and I think adding a generalized version would be useful for others.

**Comment by chrissyhroberts** [comment]
Yes so in the read me file I talk about 

A) enrolment, B) verification and C) identification

The existing features cover enrolment and verification. This function would be identification. 

There's already an example of how to do this with the CLI but an on device identification would be great. My guess is that it would probably be pretty unfeasible at scale, i.e. Because android devices have limited multi threading and the database of references templates would be big. 

Obvious solution would be to partition the problem to reduce the set of references. This would be like using a choice filter 

Let's say you start with a db of 10000 reference templates. 

You aren't going to be able to compare index template to 10000 in rational time. 

So you filter on region, gender, age group and whatever, each time reducing the reference set. After a few parameters are used you'll be down to tens of references, then you run the comparison. 

Issue will be when people move district or change age because they got older and so on. 

Very important feature

**Comment by tobiasmcnulty** [comment]
@chrissyhroberts Thanks for the quick response. Just to clarify, our CLI tool is more aimed at a deduplication context rather than identification per se.

That said, identification is an interesting thought and could certainly be accommodated in the Android app with much of the same code.

We observed around 240,000 matches per second on server hardware, so while the match rate would certainly be less on an Android device, it could still be quite tolerable, depending on the population size of course.

**Comment by chrissyhroberts** [comment]
@tobiasmcnulty I understand need for a deduplication tool, but questions arise in my head about how to know which of two, three or more replicate records to keep if the data content (i.e. name, place, etc) differs between them. Would need some careful thought depending on the context in which it is being used. If someone was recruited twice for a clinical trial, there would be probably relatively limited potential for two records to differ substantially, like maybe just spelling of name differs; so merging or review and deletion of duplicate might be easy enough. But if there was a reason why someone might want to register twice under different personas - i.e. to carry out fraud, then you might have a harder time rationalising the ONE record to keep. 

For identification on device, I’d like to see how multithreading matches on Android performs in the real world and also how you could extract a set of templates for references when screening. At the moment you have to push specific number of templates, but this would need more flexibility to pull a vector of values and test them all.

**Comment by tobiasmcnulty** [comment]
@chrissyhroberts The questions you raise indeed represent a hard problem with an answer that depends largely on the context. To be clear, I don't think the Keppel CLI should set out to solve this problem. It would merely be a matching engine to help run SourceAFIS efficiently at scale. While the CLI already runs SourceAFIS as a matching engine, it's too slow for anything except the smallest of datasets, due to the need to spin up a new Java process and re-parse the templates for every match. Anecdotally, Kafis improves the matching performance by roughly 5,000 times (24 matches per second currently vs 120,000 per second or higher with the improvements, with both tests on a developer laptop). As such, we thought the implementation might be more generally useful as @seadowg described. Of course, we are happy to maintain Kafis separately if you prefer.

For identification, I'm not sure how you would load the full set of templates into Keppel. You would definitely want to parse them only once and keep them in memory for each identification attempt, otherwise I can almost guarantee it would be too slow to be practical. While it might share some of the same code, it feels sufficiently different as a use case to probably merit its own issue. I hope that helps.

**Comment by chrissyhroberts** [comment]
very happy to roll Kafis in to the main repo as it seems to be a big upgrade
